### DISCUSSION TITLE
[help] Group iteration does not expand branches as expected

### DISCUSSION THREAD

**Discussion by lunafazio** [ropensci/targets#1226] [Help] [answered]
### Help

- [x] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi, apparently using `iteration = "group"` is ignored when the pattern is downstream from another pattern? Here's a simple example with some commentary explaining why it would make sense to do things this way:

```r
library(dplyr)
library(targets)
library(tidyr)

list(
  tar_target(
    nested_df,
    tibble(
      id = 1:3,
      nest = rep(list(letters[1:3]), 3)
    )
  ),
  tar_target(
    nested_row,
    # do something expensive per nested row
    nested_df %>% mutate(res = paste0("res", id)),
    pattern = map(nested_df)
  ),
  tar_target(
    unnested_row,
    # expensive result gets replicated
    unnest(nested_row, nest),
    pattern = map(nested_row)
  ),
  tar_target(
    grouped_row,
    # introduce groups
    unnested_row %>%
      group_by(id,nest) %>%
      tar_group(),
    pattern = map(unnested_row),
    iteration = "group"
  ),
  tar_target(
    over_rows,
    # try to do further expensive things per separate row
    # (not actually separate -> risk of e.g. memory overflow)
    rep(
      numeric(10^9),
      length(grouped_row$tar_group)
    ),
    pattern = map(grouped_row),
    iteration = "list"
  )
)
```

I don't necessarily expect this to work the way I want it to. Another sensible outcome (though undesirable for me) would be to get multiple rows all with the same `tar_group` value, but as it currently stands, it just returns all the rows as if no grouping had been requested, which feels like it's just silently failing to do anything.

I had a similar situation in #1113, where Will suggested that cheap operations could be performed before the actual pipeline and arranged as desired into a dataframe. That worked at the time, but no longer applies here as some of the steps I want to unnest over are rather complicated.

I've also looked over #348, #554, #971. It seems like his recommendation for these situations is to use static branching, but I'd like to know if there is some other alternative as this results in a network viz with redundant clutter. I've considered saving the actual big object as a side effect and passing along the filename so I can perform the splits I want on a pattern-free target but it feels a bit hacky (and doing it directly is not an option since it would involve loading all the large objects in memory at once).

**Comment by wlandau** [comment, marked as answer]
`iteration = "group"` is meant for non-dynamic targets. It controls how a single data frame target is split into row groups for downstream dynamic targets. Because `grouped_row` is already dynamic with `pattern = map(grouped_row)`, the target is already split into branches, and `iteration = "group"` no longer makes sense. So I recommend removing  `pattern = map(grouped_row)` from `grouped_row` and keeping `iteration = "group"`. As long as `iteration` is `"vector"` (default) for `unnested_row`, the branches of `unnested_row` will be automatically aggregated into a data frame. This should be fine in your example because things don't get demanding on memory or computation until the `over_rows` target.

**Reply by wlandau** [reply]
a33b21e7ff088daceeda9225b1eb18c700b3eec6 adds an explicit error check to make sure `iteration = "group"` and `pattern = map(...)` are not both set in the same target.

**Reply by lunafazio** [reply]
Ah, I realize now that my example didn't actually properly illustrate my point. You are right, as it stands `unnested_row` would not need to be a pattern; however, in my actual use case, the "expensive results" getting replicated are themselves large, so holding all of them in memory simultaneously would not work.

Additionally, I realised now that even if did work, it would cause these large objects to be saved twice. I suppose this could be circumvented by using a proper file target, but that's a separate issue.

Still, thank you for your answer!

**Reply by wlandau** [reply]
>  the "expensive results" getting replicated are themselves large, so holding all of them in memory simultaneously would not work.

In that case, it may help to consider an approach different than group iteration. When `targets` dynamically branches over an upstream non-dynamic target, it needs to hold that upstream target in memory somewhere in order to create the slices.
